src/optimizers/anyprecision_optimizer.py

```diff
@@ -0,0 +1,182 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+# All rights reserved.
+#
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+# AnyPrecisionAdamW: a flexible precision AdamW optimizer
+# with optional Kahan summation for high precision weight updates.
+# Allows direct control over momentum, variance and auxiliary compensation
+# buffer dtypes.
+# Optional Kahan summation is used to offset precision reduction for
+# the weight updates. This allows full training in BFloat16 (equal or
+# better than FP32 results in many cases) due to high precision weight upates.
+
+import torch
+from torch.optim.optimizer import Optimizer
+
+
+class AnyPrecisionAdamW(Optimizer):
+    def __init__(
+        self,
+        params,
+        lr=1e-3,
+        betas=(0.9, 0.999),
+        eps=1e-8,
+        weight_decay=0.0,
+        use_kahan_summation=False,
+        momentum_dtype=torch.float32,
+        variance_dtype=torch.bfloat16,
+        compensation_buffer_dtype=torch.bfloat16,
+    ):
+        """
+        Args:
+                params (iterable): iterable of parameters to optimize or dicts defining
+                    parameter groups
+                lr (float, optional): learning rate (default: 1e-3)
+                betas (Tuple[float, float], optional): coefficients used for computing
+                    running averages of gradient and its square (default: (0.9, 0.999))
+                eps (float, optional): term added to the denominator to improve
+                    numerical stability (default: 1e-8)
+                weight_decay (float, optional): weight decay coefficient (default: 1e-2)
+
+                # Any Precision specific
+                use_kahan_summation = creates auxiliary buffer to ensure high precision
+                model param updates (default: False)
+                momentum_dtype = dtype for momentum  (default: BFloat32)
+                variance_dtype = dtype for uncentered variance (default: BFloat16)
+                compensation_buffer_dtype  = dtype for Kahan summation
+                                             buffer (default: BFloat16)
```

Comment:
why is the default bf16 and not fp16? won't it be more precise than bf16? Thanks!

Reply:
Hi @stas00  - BF16 is a drop in replacement for FP32 b/c it has the same dynamic range. 
FP16 requires scaling due to the lower than FP32 dynamic range. 

Technically FP16 does have greater precision, but the fact that the range of it is so poor means you need to rescale in order to effectively use. (hence all the grad_scaler you see, and the reason many have given up on FP16 for LLM training). 

Thus, I use BF16 here since it is a drop in replacement (though yes with lower precision than FP16 and FP32) for FP32, and no rescaling (aka guessing) needed. 
Hope that helps!

Reply:
Thank you for the follow up, Less

But the compensation buffer is to save the error that didn't get added up. If the error is smaller than what bf16 can handle, aren't we getting "an error" on "an error" here? I mean the compensation buffer could get zero'ed out as well in such cases.

And of course you're right that the compensation buffer is also used in this code to add the big numbers - fp16 could easily overflow - yes. Perhaps only the error should be stored in a higher precision format and be added and not be added to?

Would it be a "safer" default to use fp32? And let the user decide if they want the error to be not fully carried over.

On the other hand if the grads are in bfloat16 then there will be no loss if the compensation buffer is in bf16 I think.

Reply:
Hi Stas,
I see what you mean here - it is possible that fp32 might improve the precision for the comp buffer.  The flip side is that of course increases the memory, and and we also have overhead from up and downcasting. 
Originally I was focused on creating a pure bfloat16 pipeline (i.e. model, optimizer, etc) all in pure bfloat16 so that was really the main focus.
With that accomplished, we could review additional optimizations like customizing the comp buffer.  Note that I had tried using stochastic rounding to remove the need for the buffer, but that did not work as well and so dropped that. 
I also did some work with int8 quantized (dynamic block quantization) as the buffer, but seemed to be problematic when used for fine tuning. 

I could test out some different alternatives to try and compare impact - do you have one or two recommended models/training examples that would be representative of your users most typical use cases?
That way I can setup and we can plug in different comp options and compare the tradeoffs.

Reply:
Thank you for sharing the background story, Less

I don't have any specific examples at the moment.

But I have just trained from scratch OPT-1.3b in fp16/fp32/bf16 - mixed precision for non-fp32 - using the exact same setup otherwise. As you can see bf16 gives a much much slower loss curve.

![snapshot_69](https://user-images.githubusercontent.com/10676103/214652979-247fdebb-2c52-4e67-9428-6549ba72fc02.png)

In order to catch up with fp16/fp32 I need to feed 10x more batchsize to bf16 setup.

This of course isn't the same situation, but it's quite telling how bf16, while really helpful at avoiding instabilities when training huge models, is a much much slower training regime otherwise. Hence the concerns of having it as the default.

Reply:
Hi Stas,
Thanks for the updates.
Can you share more info on the above loss curve - when you say you trained 'in bf16', does that mean you trained with pure bf16 or just the model weights in bf16 or ? 
If full bf16, did you use anyprecision and kahan summation with it?  
Based on the curve above and the comments you need 10x more data, you are hitting some level of weight stagnation, which is what Kahan summation is designed to avoid.
In my testing (fine tuning though) results were very similar with mixed precision fp32 and anyprecision bf16, except bf16 trained faster.
Re: instability of fp16 - that really begins to show at 5B+ model size from what I've seen and is basically unusable without substantial restarts for 20B+ models.

Reply:
> Hi Stas, Thanks for the updates. Can you share more info on the above loss curve - when you say you trained 'in bf16', does that mean you trained with pure bf16 or just the model weights in bf16 or ? 

`amp.autocast(bf16)` and normal adamw (not yet anyprecision)

> If full bf16, did you use anyprecision and kahan summation with it? Based on the curve above and the comments you need 10x more data, you are hitting some level of weight stagnation, which is what Kahan summation is designed to avoid. In my testing (fine tuning though) results were very similar with mixed precision fp32 and anyprecision bf16, except bf16 trained faster. 

I'd want to try that next then.

Could you please remind me which pytorch already has anyprecision? 1.13 or nightly?

> Re: instability of fp16 - that really begins to show at 5B+ model size from what I've seen and is basically unusable without substantial restarts for 20B+ models.

Yes, that's about right!

Reply:
> you are hitting some level of weight stagnation,

do you by chance have some pointers to learn about that subject matter?

my straight intuition is that bf16's low precision simply doesn't provide enough signal to learn fast with smaller batches.

Reply:
ok, so if I'm not building from scratch I have to use pt-1.12 for prebuilt `torchdistx`

Reply:
ah, no, none of the released versions have `torchdistx.optimizers` :( the nightly isn't nightly either - it's `torchdistx-0.2.0.dev20220704+cu116`

Reply:
Hi Stas,
That's too much overhead for testing imo - can you just drop the .py file and then you can run whatever PT you are using.  (Torchdistx was linked via building b/c of MetaTensor, but FakeTensor has surpassed it and so that's now an unwelcome tie down).

Reply:
You can just drop in the file and then init the optimizer and make sure to set:
use_kahan_summation=True,
momentum_dtype=torch.bfloat16,
variance_dtype=torch.bfloat16,
compensation_buffer_dtype=torch.bfloat16,

So that you have a pure BF16 setup ready to go.
Please let me know if any questions - I'll talk w Rohan re: torchdistx dependency as that's a drag atm.

Reply:
yup, Icame to the same conclusion - I just did:

```
mkdir -p $CONDA_PREFIX/lib/python3.8/site-packages/torchdistx/optimizers
wget https://raw.githubusercontent.com/pytorch/torchdistx/main/src/python/torchdistx/optimizers/anyprecision_optimizer.py -O $CONDA_PREFIX/lib/python3.8/site-packages/torchdistx/optimizers/__init__.py
```

so the model itself is in bf16 too - no AMP, correct?

Reply:
re: install - perhaps this can be upstreamed to pytorch-core for 1.14 release? it's just a single standalone python file.

Reply:
Also yes, just move the whole model to BF16:

~~~~
if cfg.model_in_bf16:
        model.to(torch.bfloat16)
~~~~
and then use AnyPrecision with it, setting the variance and state to BF16 and Kahan to True.
That will give you 100% pure BFloat16 training, and let's see how it does relative to the earlier BF16 run.

Reply:
Hi Stas,
Actually if you want to share your setup for the opt run above, I can likely run it here and test out Anyprecision as well as a couple tweaks to the comp buffer.

Reply:
Thank you, Less, 

I had to hack the HF Trainer a bit to make it work, so WIP is here, including full instructions:
https://github.com/huggingface/transformers/pull/21312

Please let me know if you run into any issues.

I have it loaded in a slurm job as well, might need to wait for it to yield - but testing with a larger local dataset.

Reply:
Though I think something might be wrong with my integration, since I don't see any memory usage reduction. Will check closer tomorrow.

Reply:
Hi Stas,
Thanks for the instructions. I got setup on my a10 but I guess this trainer is not using fsdp?  I ran out of gpu ram vs 1.3B readily fits with FSDP. 
I'll have to setup on a larger server then. (is this DDP underneath this trainer?)
Also talked to Rohan so he may try to move anyprecision up to PT-D since there is no dependency on metatensor.

Reply:
oh, I am using 80GB A100, yes, 24GB won't be enough. You probably could use fsdp, I use deepspeed for zero.

Reply:
Less, the other approach is to just use an even smaller model so that it'd fit into your setup. e.g. use 1/8th of it?
e.g. could train [opt-125m](https://huggingface.co/facebook/opt-125m) instead of opt-1.3b

Reply:
and so the goal is to show that bf16-pure/anyprecision(bf16) trains as fast as fp16/amp/adamw - which currently isn't the case (at least with opt-1.3)

Reply:
I tried on a larger server but still each card is A10 23GB...seems you are expecting A100s and 40GB?  Not sure what the trainer is doing exactly but this would easily run with FSDP. 
Can you clarify the expected hardware to run this?

Reply:
You can see the details here: https://github.com/pytorch/torchdistx/pull/52#issuecomment-1405367731

As I suggested let's switch to opt-125m instead

Reply:
I updated the instructions here: https://github.com/huggingface/transformers/pull/21312 to include 125m - 10x smaller - should fit into 24GB card no problem.

Reply:
Hi @lessw2020 - we are still interested in your suggestion that you were able to train pure bf16 faster than fp16/amp.

how do I reproduce it?